### PR TITLE
slot2ssa: Fix spurious φᶜ edges

### LINKED ISSUE #51159: Unreachable reached due to TypedSlot removal
MWE:

```julia
function f()
    result = false
    for i in 3
        result = try
        catch
            continue
        end
    end
    result
end
f()
```

```
$ manyjulias master wip.jl
[ Info: Translated requested revision master to commit ec2f1d3c0b2a37b5acafeaab2454fedbd83a31ab
Unreachable reached at 0x7f39379632a6

[2123194] signal (4.2): Illegal instruction
in expression starting at /home/maleadt/julia/wip.jl:16
f at /home/maleadt/julia/wip.jl:9
```

Bisected to #50943

As spotted on PkgEval during BeaData.jl tests: https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2023-09/01/BeaData.primary.log (but code mostly comes from Retry.jl).

### PR DESCRIPTION
**Pull request by topolarity**
The unreachable here seems to be caused by the fact that (as of #50943) we re-use a more narrow type from Inference that correctly ignores these edges, but when inserting the `φᶜ` node in `slot2reg` we were including extra edges that never get exercised at runtime.

I'm not sure _why_ this causes us to hit an unreachable, since the narrow type from inference is technically still valid (the catch block will never observe these spurious assignments at runtime), but this seems to fix the issue and anyway improves the quality of the IRCode generated by `slot2ssa`.

Resolves #51159

### CODE REVIEW

**Comment by vtjnash** [comment]
`φᶜ` assignment happens semantically at the earlier `ɣ`, just as `φ` assignment happens at the preceding edge. The actual `φᶜ` node is just a placeholder for dominator analysis and ssa numbering.

File: test/compiler/irpasses.jl [hunk 1]
```diff
@@ -4,7 +4,7 @@ using Test
 using Base.Meta
 import Core:
```
Comment (aviatesk) [inline test/compiler/irpasses.jl:5]:
```suggestion
using Core.IR
```
would be cleaner.

File: test/compiler/irpasses.jl [hunk 2]
```diff
@@ -1491,3 +1491,31 @@ function call_big_dead_throw_catch()
...
+    for i = 1:length(ir.stmts)
+        expr = ir.stmts[i][:stmt]
+        if isa(expr, PhiCNode)
+            # The φᶜ should only observe the value of `result` at the try-catch :enter
+            # (from the `result = false` assignment), since `result = x` assignment is
+            # dominated by an Expr(:leave).
+            @test length(expr.values) == 1
+        end
+    end
+end
```
Comment (aviatesk) [inline test/compiler/irpasses.jl:1521]:
```suggestion
end
@test isnothing(f_with_early_try_catch_exit())
```

**Review by aviatesk — APPROVED** [review]

**Comment by topolarity** [comment]
> `φᶜ` assignment happens semantically at the earlier `ɣ`

Do you mean that was happening even though the ɣ was outside of any try-catch block?

File: test/compiler/irpasses.jl (hunk 1 above)
Comment (topolarity) [reply test/compiler/irpasses.jl:5]: Hmm, I really appreciate explicit imports, personally.

Easily >50% of my confusion reading new Julia code is figuring out where a name came from, but in this case it's pretty obvious so I'm happy to make the change.

File: test/compiler/irpasses.jl (hunk 2 above)
Comment (topolarity) [reply test/compiler/irpasses.jl:1521]: Good idea!

**Review by aviatesk — APPROVED** [review]

File: test/compiler/irpasses.jl [hunk 3]
```diff
@@ -1491,3 +1489,33 @@ function call_big_dead_throw_catch()
     end
     return 4
 end
+
+# Issue #51159 - Unreachable reached in try-catch block
+function f_with_early_try_catch_exit()
+    result = false
+    for i in 3
```
Comment (IanButterworth) [inline test/compiler/irpasses.jl:1496]: Is this meant to be `for i in 1:3`?

**Pull request merged**